src/main/java/de/maxhenkel/reap/TreeEvents.java

```diff
@@ -59,97 +63,104 @@ public static boolean canHarvest(BlockPos pos, Player player, Level world, ItemS
+        final Block wantedBlock = world.getBlockState(start).getBlock();
+        boolean abort = false;
+        final LinkedList<BlockPos> result = new LinkedList<>();
+        final Set<BlockPos> visited = new HashSet<>();
+        final LinkedList<BlockPos> queue = new LinkedList<>();
         int maxHarvestingCount = Main.SERVER_CONFIG.treeHarvestMaxCount.get();
-        if (positions.size() >= maxHarvestingCount) {
-            return;
-        }
-        List<BlockPos> posList = new ArrayList<>();
-        for (int x = -1; x <= 1; x++) {
-            for (int y = -1; y <= 1; y++) {
-                for (int z = -1; z <= 1; z++) {
-                    BlockPos p = pos.offset(x, y, z);
-                    if (isLog(world, p)) {
-                        if (positions.size() <= maxHarvestingCount) {
-                            if (positions.add(p)) {
-                                posList.add(p);
+        queue.push(start);
+
+        while (!queue.isEmpty()) {
+            final BlockPos center = queue.pop();
+            final int x0 = center.getX();
+            final int y0 = center.getY();
+            final int z0 = center.getZ();
+            for (int z = z0 - 1; z <= z0 + 1 && !abort; ++z) {
+                for (int y = y0 - 1; y <= y0 + 1 && !abort; ++y) {
+                    for (int x = x0 - 1; x <= x0 + 1 && !abort; ++x) {
+                        final BlockPos pos = new BlockPos(x, y, z);
+                        final BlockState bs = world.getBlockState(pos);
+                        if ((bs.isAir() || !visited.add(pos))) {
+                            continue;
+                        }
+                        final boolean isRightBlock = bs.is(wantedBlock);
+                        if (isRightBlock) {
+                            result.add(pos);
+                            if (queue.size() > maxHarvestingCount) {
+                                abort = true;
+                                break;
                             }
-                        } else {
-                            return;
+                            queue.push(pos);
+                        }
+                        if (action != null) {
+                            abort = !action.onBlock(pos, bs, isRightBlock);
                         }
                     }
                 }
             }
         }
-
-        for (BlockPos p : posList) {
-            collectLogs(world, p, positions);
-        }
+        return !abort ? result : new LinkedList<>();
     }
 
-    private static boolean isLog(Level world, BlockPos pos) {
-        BlockState b = world.getBlockState(pos);
-        return Main.SERVER_CONFIG.logTypes.stream().anyMatch(tag -> tag.contains(b.getBlock()));
-    }
-
-    private static boolean isGround(Level world, BlockPos pos) {
-        BlockState b = world.getBlockState(pos);
-        return Main.SERVER_CONFIG.groundTypes.stream().anyMatch(tag -> tag.contains(b.getBlock()));
```

Comment:
Why did you remove the ground check?

Reply:
Because you won't be able to cut trees that are left in air. or when the tool's durability is gone and there are logs left in air.

Reply:
Yeah its meant to work like this. You should only be able to fell a tree if its on the ground and you break the bottom log.

Reply:
This is a little strange imo, but I get that. 
I think this is ready for review.